example/cli/odp_cli.c

```diff
@@ -0,0 +1,145 @@
+/* Copyright (c) 2021, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/*
+ * ODP CLI Helper Example
+ *
+ * This example shows how to start and stop ODP CLI using the CLI helper
+ * API functions. This example application can also be used to try out
+ * the CLI by connecting to a running application with a telnet client.
+ */
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+#include <odp/helper/cli.h>
+
+#include <stdio.h>
+#include <stdint.h>
+
+typedef struct {
+	int num;
+	char *addr;
+	uint16_t port;
+} options_t;
+
+static void usage(const char *prog)
+{
+	printf("\n"
+	       "Usage: %s [options]\n"
+	       "\n"
+	       "OPTIONS:\n"
+	       "  -n, --num <sec>         Keep CLI open for <sec> seconds. (default 0)\n"
+	       "  -a, --address <addr>    Bind listening socket to IP address <addr>.\n"
+	       "  -p, --port <port>       Bind listening socket to port <port>.\n"
+	       "\n"
+	       "ODP helper defaults are used for address and port, if the options are\n"
+	       "not given.\n"
+	       "\n",
+	       prog);
+}
+
+static void parse_args(int argc, char *argv[], options_t *opt)
+{
+	static const struct option longopts[] = {
+		{ "num", required_argument, NULL, 'n' },
+		{ "address", required_argument, NULL, 'a' },
+		{ "port", required_argument, NULL, 'p' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+n:a:p:h";
+
+	while (1) {
+		int c = getopt_long(argc, argv, shortopts, longopts, NULL);
+
+		if (c == -1)
+			break; /* No more options */
+
+		switch (c) {
+		case 'n':
+			opt->num = atoi(optarg);
+			break;
+		case 'a':
+			opt->addr = optarg;
+			break;
+		case 'p':
+			opt->port = atoi(optarg);
+			break;
+		default:
+			usage(argv[0]);
+			exit(EXIT_SUCCESS);
+			break;
+		}
+	}
+
+	optind = 1; /* reset 'extern optind' from the getopt lib */
+}
+
+int main(int argc, char *argv[])
+{
+	odp_instance_t inst;
+	options_t opt;
+	int i;
+
+	memset(&opt, 0, sizeof(opt));
+	parse_args(argc, argv, &opt);
+
+	/* Initialize ODP. */
+
+	if (odp_init_global(&inst, NULL, NULL)) {
+		printf("Global init failed.\n");
+		return -1;
```

Comment:
Please use `ODPH_ERR()` macro for error messages and exit with `exit(EXIT_FAILURE)`.

Reply:
Again, copied this from example/hello. I'll change it.